src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -1088,6 +917,368 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+                }
+
+                rit->second.remove_participant(guid.guidPrefix);
+            }
+        }
+    }
+
+    // remove this writer from its participant
+    auto pit = participants_.find(guid.guidPrefix);
+    if (pit == participants_.end())
+    {
+        logError(DISCOVERY_DATABASE,
+                "Unexisting participant " << guid.guidPrefix << " for writer: " << guid);
+        return;
+    }
+
+    pit->second.remove_writer(guid);
+}
+
+void DiscoveryDataBase::unmatch_reader_(const eprosima::fastrtps::rtps::GUID_t& guid)
+{
+    logInfo(DISCOVERY_DATABASE, "unmatching reader: " << guid);
+
+    auto rit = readers_.find(guid);
+    if (rit == readers_.end())
+    {
+        logWarning(DISCOVERY_DATABASE,
+                "Attempting to unmatch an unexisting reader: " << guid);
+        return;
+    }
+
+    // get reader topic
+    std::string topic = rit->second.topic();
+
+    // remove it from reader by topic
+    remove_reader_from_topic_(guid, topic);
+
+    // it there are more than one reader in this topic in the same participant we do not unmatch the endpoints
+    if (repeated_reader_topic_(guid.guidPrefix, topic))
+    {
+        // for each writer in same topic make not relevant. It could be none in writers
+        auto tit = writers_by_topic_.find(topic);
+        if (tit != writers_by_topic_.end())
+        {
+            for (auto writer : tit->second)
+            {
+                auto wit = writers_.find(writer);
+                if (wit == writers_.end())
+                {
+                    logWarning(DISCOVERY_DATABASE,
+                            "Unexisting writer " << writer << " in topic: " << topic);
+                }
+
+                wit->second.remove_participant(guid.guidPrefix);
+            }
+        }
+    }
+
+    // remove this reader from its participant
+    auto pit = participants_.find(guid.guidPrefix);
+    if (pit == participants_.end())
+    {
+        logError(DISCOVERY_DATABASE,
+                "Unexisting participant " << guid.guidPrefix << " for reader: " << guid);
+        return;
+    }
+    pit->second.remove_reader(guid);
+}
+
+
+bool DiscoveryDataBase::repeated_writer_topic_(const eprosima::fastrtps::rtps::GuidPrefix_t& participant,
+        const std::string& topic_name)
+{
+    int count = 0;
+
+    auto pit = participants_.find(participant);
+    if (pit == participants_.end())
+    {
+        logWarning(DISCOVERY_DATABASE,
+                "Checking repeated writer topics in an unexisting participant: " << participant);
+        return false;
+    }
+
+    for (auto writer_guid : pit->second.writers())
+    {
+        auto wit = writers_.find(writer_guid);
+        if (wit == writers_.end())
+        {
+            logWarning(DISCOVERY_DATABASE,
+                    "writer missing: " << writer_guid);
+        }
+
+        if (wit->second.topic() == topic_name)
+        {
+            ++count;
```

Comment:
There is no need to finish the loop. 
At the moment you do `++count` twice you already know that the count will be higher that 1.
However I don't know if it is more efficient to check count on each iteration or finish the loop.
I'd do the first.

Reply:
Good catch